File: game_collection.rb
# game_collection.rb
require 'json'
require 'date'
require 'csv'

class Game
  attr_accessor :id, :title, :genre, :platform, :year, :completed, :rating, :notes, :added_date

  def initialize(id, title, genre, platform, year, completed, rating, notes = "", added_date = Date.today.to_s)
    @id = id
    @title = title
    @genre = genre
    @platform = platform
    @year = year
    @completed = completed
    @rating = rating
    @notes = notes
    @added_date = added_date
  end

  def to_h
    { id: @id, title: @title, genre: @genre, platform: @platform, year: @year,
      completed: @completed, rating: @rating, notes: @notes, added_date: @added_date }
  end

  def self.from_h(hash)
    Game.new(hash[:id], hash[:title], hash[:genre], hash[:platform], hash[:year],
             hash[:completed], hash[:rating], hash[:notes], hash[:added_date])
  end
end

class GameCollection
  attr_reader :games

  def initialize
    @games = []
    @next_id = 1
  end

  def add_game(title, genre, platform, year, completed, rating, notes = "")
    raise "Оценка должна быть от 1 до 10" unless (1..10).include?(rating)
    raise "Год должен быть от 1980 до #{Date.today.year}" unless (1980..Date.today.year).include?(year)
    raise "Название, жанр и платформа не могут быть пустыми" if title.empty? || genre.empty? || platform.empty?
    game = Game.new(@next_id, title, genre, platform, year, completed, rating, notes)
    @games << game
    @next_id += 1
    game
  end

  def find_game(id)
    @games.find { |g| g.id == id }
  end

  def edit_game(id, **kwargs)
    game = find_game(id)
    return false unless game
    kwargs.each do |key, value|
      game.send("#{key}=", value) if game.respond_to?("#{key}=")
    end
    true
  end

  def delete_game(id)
    game = find_game(id)
    return false unless game
    @games.delete(game)
    true
  end

  def search_games(query)
    q = query.downcase
    @games.select { |g| g.title.downcase.include?(q) }
  end

  def filter_by_completed(completed)
    @games.select { |g| g.completed == completed }
  end

  def filter_by_genre(genre)
    @games.select { |g| g.genre.downcase == genre.downcase }
  end

  def filter_by_platform(platform)
    @games.select { |g| g.platform.downcase == platform.downcase }
  end

  def sort_by_rating(reverse = true)
    @games.sort_by { |g| g.rating }.reverse! if reverse
    @games.sort_by { |g| g.rating }
  end

  def sort_by_title
    @games.sort_by { |g| g.title.downcase }
  end

  def stats
    total = @games.size
    completed_count = filter_by_completed(true).size
    uncompleted = total - completed_count
    ratings = filter_by_completed(true).map(&:rating)
    avg_rating = ratings.empty? ? 0 : ratings.sum.to_f / ratings.size
    platforms = Hash.new(0)
    genres = Hash.new(0)
    @games.each do |g|
      platforms[g.platform] += 1
      genres[g.genre] += 1
    end
    { total: total, completed: completed_count, uncompleted: uncompleted,
      avg_rating: avg_rating, platforms: platforms, genres: genres }
  end

  def save_to_file(filename = "games_data.json")
    data = { games: @games.map(&:to_h) }
    File.write(filename, JSON.pretty_generate(data))
  end

  def load_from_file(filename = "games_data.json")
    return unless File.exist?(filename)
    data = JSON.parse(File.read(filename), symbolize_names: true)
    @games.clear
    data[:games].each do |item|
      game = Game.from_h(item)
      @games << game
      @next_id = game.id + 1 if game.id >= @next_id
    end
  rescue JSON::ParserError
    puts "Ошибка чтения файла."
  end

  def export_csv(filename = "games_export.csv")
    CSV.open(filename, "w", col_sep: ";") do |csv|
      csv << ["ID", "Название", "Жанр", "Платформа", "Год", "Пройдена", "Оценка", "Заметки", "Дата добавления"]
      @games.each do |g|
        csv << [g.id, g.title, g.genre, g.platform, g.year, g.completed ? "Да" : "Нет", g.rating, g.notes, g.added_date]
      end
    end
  end

  def import_csv(filename = "games_export.csv")
    unless File.exist?(filename)
      raise "Файл не найден"
    end
    CSV.foreach(filename, headers: true, col_sep: ";") do |row|
      begin
        add_game(
          title: row["Название"],
          genre: row["Жанр"],
          platform: row["Платформа"],
          year: row["Год"].to_i,
          completed: row["Пройдена"] == "Да",
          rating: row["Оценка"].to_i,
          notes: row["Заметки"]
        )
      rescue => e
        puts "Ошибка импорта строки: #{e}"
      end
    end
  end
end

def print_game(game)
  status = game.completed ? "✅ Пройдена" : "⏳ Не пройдена"
  puts "##{game.id} - #{game.title} (#{game.year})"
  puts "   Жанр: #{game.genre}, Платформа: #{game.platform}"
  puts "   #{status}, Оценка: #{game.rating}/10"
  puts "   Заметки: #{game.notes}" unless game.notes.empty?
  puts "   Добавлена: #{game.added_date}"
end

def main
  collection = GameCollection.new
  collection.load_from_file

  loop do
    puts "\n===== КОЛЛЕКЦИЯ ИГР (Ruby) ====="
    puts "1. Добавить игру"
    puts "2. Показать все игры"
    puts "3. Показать пройденные игры"
    puts "4. Показать непройденные игры"
    puts "5. Найти игры по названию"
    puts "6. Сортировать по оценке (по убыванию)"
    puts "7. Сортировать по названию"
    puts "8. Редактировать игру"
    puts "9. Удалить игру"
    puts "10. Показать статистику"
    puts "11. Сохранить в файл"
    puts "12. Загрузить из файла"
    puts "13. Экспорт в CSV"
    puts "14. Импорт из CSV"
    puts "0. Выход"
    print "Выберите действие: "
    choice = gets.chomp

    case choice
    when "0"
      break
    when "1"
      print "Название: "
      title = gets.chomp
      next if title.empty?
      print "Жанр: "
      genre = gets.chomp
      next if genre.empty?
      print "Платформа: "
      platform = gets.chomp
      next if platform.empty?
      print "Год выпуска: "
      year = gets.chomp.to_i
      print "Статус (1-пройдена, 0-нет): "
      completed = gets.chomp == "1"
      print "Оценка (1-10): "
      rating = gets.chomp.to_i
      print "Заметки (необязательно): "
      notes = gets.chomp
      begin
        game = collection.add_game(title, genre, platform, year, completed, rating, notes)
        puts "Игра добавлена с ID #{game.id}"
      rescue => e
        puts "Ошибка: #{e.message}"
      end
    when "2"
      if collection.games.empty?
        puts "Нет игр."
      else
        collection.games.each { |g| print_game(g) }
      end
    when "3"
      games = collection.filter_by_completed(true)
      if games.empty?
        puts "Нет пройденных игр."
      else
        games.each { |g| print_game(g) }
      end
    when "4"
      games = collection.filter_by_completed(false)
      if games.empty?
        puts "Нет непройденных игр."
      else
        games.each { |g| print_game(g) }
      end
    when "5"
      print "Введите часть названия: "
      query = gets.chomp
      results = collection.search_games(query)
      if results.empty?
        puts "Игры не найдены."
      else
        results.each { |g| print_game(g) }
      end
    when "6"
      sorted = collection.sort_by_rating(reverse: true)
      if sorted.empty?
        puts "Нет игр."
      else
        sorted.each { |g| print_game(g) }
      end
    when "7"
      sorted = collection.sort_by_title
      if sorted.empty?
        puts "Нет игр."
      else
        sorted.each { |g| print_game(g) }
      end
    when "8"
      print "Введите ID игры для редактирования: "
      id = gets.chomp.to_i
      game = collection.find_game(id)
      unless game
        puts "Игра не найдена."
        next
      end
      puts "Оставьте поле пустым, чтобы не менять."
      print "Название (#{game.title}): "
      new_title = gets.chomp
      print "Жанр (#{game.genre}): "
      new_genre = gets.chomp
      print "Платформа (#{game.platform}): "
      new_platform = gets.chomp
      print "Год (#{game.year}): "
      new_year = gets.chomp
      print "Статус (1-пройдена, 0-нет) сейчас: #{game.completed ? '1' : '0'}: "
      new_completed = gets.chomp
      print "Оценка (#{game.rating}): "
      new_rating = gets.chomp
      print "Заметки (#{game.notes}): "
      new_notes = gets.chomp
      updates = {}
      updates[:title] = new_title unless new_title.empty?
      updates[:genre] = new_genre unless new_genre.empty?
      updates[:platform] = new_platform unless new_platform.empty?
      unless new_year.empty?
        updates[:year] = new_year.to_i
      end
      unless new_completed.empty?
        updates[:completed] = new_completed == "1"
      end
      unless new_rating.empty?
        updates[:rating] = new_rating.to_i
      end
      updates[:notes] = new_notes unless new_notes.empty?
      if collection.edit_game(id, **updates)
        puts "Игра обновлена."
      else
        puts "Ошибка обновления."
      end
    when "9"
      print "Введите ID игры для удаления: "
      id = gets.chomp.to_i
      if collection.delete_game(id)
        puts "Игра удалена."
      else
        puts "Игра не найдена."
      end
    when "10"
      stats = collection.stats
      puts "\n=== СТАТИСТИКА ==="
      puts "Всего игр: #{stats[:total]}"
      puts "Пройдено: #{stats[:completed]}"
      puts "Не пройдено: #{stats[:uncompleted]}"
      puts "Средняя оценка (только пройденные): #{'%.2f' % stats[:avg_rating]}"
      puts "По платформам:"
      stats[:platforms].each { |p, c| puts "  #{p}: #{c}" }
      puts "По жанрам:"
      stats[:genres].each { |g, c| puts "  #{g}: #{c}" }
    when "11"
      collection.save_to_file
      puts "Сохранено."
    when "12"
      collection.load_from_file
      puts "Загружено."
    when "13"
      collection.export_csv
      puts "Экспортировано в games_export.csv"
    when "14"
      begin
        collection.import_csv
        puts "Импортировано из games_export.csv"
      rescue => e
        puts "Ошибка импорта: #{e}"
      end
    else
      puts "Неизвестная команда."
    end
  end
end

main if __FILE__ == $0

File: game_collection.py
# game_collection.py
import json
import csv
from dataclasses import dataclass, asdict
from datetime import date
from typing import List, Optional
from pathlib import Path

@dataclass
class Game:
    id: int
    title: str
    genre: str
    platform: str
    year: int
    completed: bool
    rating: int  # 1-10
    notes: str
    added_date: str

class GameCollection:
    def __init__(self):
        self.games: List[Game] = []
        self.next_id = 1

    def add_game(self, title: str, genre: str, platform: str, year: int,
                 completed: bool, rating: int, notes: str = "") -> Game:
        if rating < 1 or rating > 10:
            raise ValueError("Оценка должна быть от 1 до 10")
        if year < 1980 or year > date.today().year:
            raise ValueError(f"Год должен быть от 1980 до {date.today().year}")
        if not title or not genre or not platform:
            raise ValueError("Название, жанр и платформа не могут быть пустыми")
        game = Game(
            id=self.next_id,
            title=title,
            genre=genre,
            platform=platform,
            year=year,
            completed=completed,
            rating=rating,
            notes=notes,
            added_date=date.today().isoformat()
        )
        self.games.append(game)
        self.next_id += 1
        return game

    def find_game(self, game_id: int) -> Optional[Game]:
        return next((g for g in self.games if g.id == game_id), None)

    def edit_game(self, game_id: int, **kwargs) -> bool:
        game = self.find_game(game_id)
        if not game:
            return False
        for key, value in kwargs.items():
            if hasattr(game, key) and value is not None:
                setattr(game, key, value)
        return True

    def delete_game(self, game_id: int) -> bool:
        game = self.find_game(game_id)
        if game:
            self.games.remove(game)
            return True
        return False

    def search_games(self, query: str) -> List[Game]:
        q = query.lower()
        return [g for g in self.games if q in g.title.lower()]

    def filter_by_completed(self, completed: bool) -> List[Game]:
        return [g for g in self.games if g.completed == completed]

    def filter_by_genre(self, genre: str) -> List[Game]:
        return [g for g in self.games if g.genre.lower() == genre.lower()]

    def filter_by_platform(self, platform: str) -> List[Game]:
        return [g for g in self.games if g.platform.lower() == platform.lower()]

    def sort_by_rating(self, reverse: bool = True) -> List[Game]:
        return sorted(self.games, key=lambda g: g.rating, reverse=reverse)

    def sort_by_title(self) -> List[Game]:
        return sorted(self.games, key=lambda g: g.title.lower())

    def get_stats(self) -> dict:
        total = len(self.games)
        completed_count = len(self.filter_by_completed(True))
        uncompleted = total - completed_count
        ratings = [g.rating for g in self.games if g.completed]
        avg_rating = sum(ratings) / len(ratings) if ratings else 0.0
        platforms = {}
        genres = {}
        for g in self.games:
            platforms[g.platform] = platforms.get(g.platform, 0) + 1
            genres[g.genre] = genres.get(g.genre, 0) + 1
        return {
            "total": total,
            "completed": completed_count,
            "uncompleted": uncompleted,
            "avg_rating": avg_rating,
            "platforms": platforms,
            "genres": genres
        }

    def save_to_file(self, filename: str = "games_data.json") -> None:
        data = {"games": [asdict(g) for g in self.games]}
        with open(filename, "w", encoding="utf-8") as f:
            json.dump(data, f, ensure_ascii=False, indent=2)

    def load_from_file(self, filename: str = "games_data.json") -> None:
        path = Path(filename)
        if not path.exists():
            return
        with open(filename, "r", encoding="utf-8") as f:
            data = json.load(f)
            self.games.clear()
            for item in data.get("games", []):
                game = Game(
                    id=item["id"],
                    title=item["title"],
                    genre=item["genre"],
                    platform=item["platform"],
                    year=item["year"],
                    completed=item["completed"],
                    rating=item["rating"],
                    notes=item.get("notes", ""),
                    added_date=item["added_date"]
                )
                self.games.append(game)
                if game.id >= self.next_id:
                    self.next_id = game.id + 1

    def export_csv(self, filename: str = "games_export.csv") -> None:
        with open(filename, "w", newline="", encoding="utf-8") as f:
            writer = csv.writer(f, delimiter=";")
            writer.writerow(["ID", "Название", "Жанр", "Платформа", "Год", "Пройдена", "Оценка", "Заметки", "Дата добавления"])
            for g in self.games:
                writer.writerow([g.id, g.title, g.genre, g.platform, g.year,
                                 "Да" if g.completed else "Нет", g.rating, g.notes, g.added_date])

    def import_csv(self, filename: str = "games_export.csv") -> None:
        path = Path(filename)
        if not path.exists():
            raise FileNotFoundError("Файл не найден")
        with open(filename, "r", encoding="utf-8") as f:
            reader = csv.DictReader(f, delimiter=";")
            for row in reader:
                try:
                    self.add_game(
                        title=row["Название"],
                        genre=row["Жанр"],
                        platform=row["Платформа"],
                        year=int(row["Год"]),
                        completed=row["Пройдена"] == "Да",
                        rating=int(row["Оценка"]),
                        notes=row["Заметки"]
                    )
                except Exception as e:
                    print(f"Ошибка импорта строки: {e}")

def print_game(game: Game) -> None:
    status = "✅ Пройдена" if game.completed else "⏳ Не пройдена"
    print(f"#{game.id} - {game.title} ({game.year})")
    print(f"   Жанр: {game.genre}, Платформа: {game.platform}")
    print(f"   {status}, Оценка: {game.rating}/10")
    if game.notes:
        print(f"   Заметки: {game.notes}")
    print(f"   Добавлена: {game.added_date}")

def main():
    collection = GameCollection()
    collection.load_from_file()

    while True:
        print("\n===== КОЛЛЕКЦИЯ ИГР (Python) =====")
        print("1. Добавить игру")
        print("2. Показать все игры")
        print("3. Показать пройденные игры")
        print("4. Показать непройденные игры")
        print("5. Найти игры по названию")
        print("6. Сортировать по оценке (по убыванию)")
        print("7. Сортировать по названию")
        print("8. Редактировать игру")
        print("9. Удалить игру")
        print("10. Показать статистику")
        print("11. Сохранить в файл")
        print("12. Загрузить из файла")
        print("13. Экспорт в CSV")
        print("14. Импорт из CSV")
        print("0. Выход")
        choice = input("Выберите действие: ").strip()

        if choice == "0":
            break
        elif choice == "1":
            title = input("Название: ").strip()
            if not title:
                print("Название не может быть пустым.")
                continue
            genre = input("Жанр: ").strip()
            if not genre:
                print("Жанр не может быть пустым.")
                continue
            platform = input("Платформа: ").strip()
            if not platform:
                print("Платформа не может быть пустой.")
                continue
            try:
                year = int(input("Год выпуска: ").strip())
            except ValueError:
                print("Введите число.")
                continue
            completed_input = input("Статус (1-пройдена, 0-нет): ").strip()
            completed = completed_input == "1"
            try:
                rating = int(input("Оценка (1-10): ").strip())
            except ValueError:
                rating = 0
            notes = input("Заметки (необязательно): ").strip()
            try:
                game = collection.add_game(title, genre, platform, year, completed, rating, notes)
                print(f"Игра добавлена с ID {game.id}")
            except Exception as e:
                print("Ошибка:", e)
        elif choice == "2":
            if not collection.games:
                print("Нет игр.")
            else:
                for g in collection.games:
                    print_game(g)
        elif choice == "3":
            games = collection.filter_by_completed(True)
            if not games:
                print("Нет пройденных игр.")
            else:
                for g in games:
                    print_game(g)
        elif choice == "4":
            games = collection.filter_by_completed(False)
            if not games:
                print("Нет непройденных игр.")
            else:
                for g in games:
                    print_game(g)
        elif choice == "5":
            query = input("Введите часть названия: ").strip()
            if not query:
                print("Введите текст.")
                continue
            results = collection.search_games(query)
            if not results:
                print("Игры не найдены.")
            else:
                for g in results:
                    print_game(g)
        elif choice == "6":
            sorted_games = collection.sort_by_rating(reverse=True)
            if not sorted_games:
                print("Нет игр.")
            else:
                for g in sorted_games:
                    print_game(g)
        elif choice == "7":
            sorted_games = collection.sort_by_title()
            if not sorted_games:
                print("Нет игр.")
            else:
                for g in sorted_games:
                    print_game(g)
        elif choice == "8":
            try:
                gid = int(input("Введите ID игры для редактирования: ").strip())
            except ValueError:
                print("Некорректный ID.")
                continue
            game = collection.find_game(gid)
            if not game:
                print("Игра не найдена.")
                continue
            print("Оставьте поле пустым, чтобы не менять.")
            new_title = input(f"Название ({game.title}): ").strip()
            new_genre = input(f"Жанр ({game.genre}): ").strip()
            new_platform = input(f"Платформа ({game.platform}): ").strip()
            new_year = input(f"Год ({game.year}): ").strip()
            new_completed = input(f"Статус (1-пройдена, 0-нет) сейчас: {'1' if game.completed else '0'}: ").strip()
            new_rating = input(f"Оценка ({game.rating}): ").strip()
            new_notes = input(f"Заметки ({game.notes}): ").strip()
            updates = {}
            if new_title: updates["title"] = new_title
            if new_genre: updates["genre"] = new_genre
            if new_platform: updates["platform"] = new_platform
            if new_year:
                try:
                    updates["year"] = int(new_year)
                except ValueError:
                    print("Год должен быть числом, пропускаем.")
            if new_completed: updates["completed"] = new_completed == "1"
            if new_rating:
                try:
                    updates["rating"] = int(new_rating)
                except ValueError:
                    print("Оценка должна быть числом, пропускаем.")
            if new_notes: updates["notes"] = new_notes
            if collection.edit_game(gid, **updates):
                print("Игра обновлена.")
            else:
                print("Ошибка обновления.")
        elif choice == "9":
            try:
                gid = int(input("Введите ID игры для удаления: ").strip())
            except ValueError:
                print("Некорректный ID.")
                continue
            if collection.delete_game(gid):
                print("Игра удалена.")
            else:
                print("Игра не найдена.")
        elif choice == "10":
            stats = collection.get_stats()
            print("\n=== СТАТИСТИКА ===")
            print(f"Всего игр: {stats['total']}")
            print(f"Пройдено: {stats['completed']}")
            print(f"Не пройдено: {stats['uncompleted']}")
            print(f"Средняя оценка (только пройденные): {stats['avg_rating']:.2f}")
            print("По платформам:")
            for p, c in stats['platforms'].items():
                print(f"  {p}: {c}")
            print("По жанрам:")
            for g, c in stats['genres'].items():
                print(f"  {g}: {c}")
        elif choice == "11":
            collection.save_to_file()
            print("Сохранено.")
        elif choice == "12":
            collection.load_from_file()
            print("Загружено.")
        elif choice == "13":
            collection.export_csv()
            print("Экспортировано в games_export.csv")
        elif choice == "14":
            try:
                collection.import_csv()
                print("Импортировано из games_export.csv")
            except FileNotFoundError:
                print("Файл games_export.csv не найден.")
            except Exception as e:
                print("Ошибка импорта:", e)
        else:
            print("Неизвестная команда.")

if __name__ == "__main__":
    main()
